src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -1,243 +1,249 @@
-        let i = ref 0 in
-        let l1 =
-          List.filter
-            (fun _ ->
-              let x = !i in
-              i := x + 1;
-              x < n)
-            l
-        in
-        i := 0;
-        let l2 =
-          List.filter
-            (fun _ ->
-              let x = !i in
-              i := x + 1;
-              x >= n)
-            l
-        in
-        Pair (from_list l1, from_list l2)
-
-  let rec to_list = function
-    | Quad (a, b, c, d) -> [ a; b; c; d ]
-    | Triple (a, b, c) -> [ a; b; c ]
-    | Pair (a, b) -> to_list a @ to_list b
-    | Elem a -> [ a ]
-
-  let to_expr ~loc t =
-    let l = to_list t in
-    let (module A) = Ast_builder.make loc in
-    List.mapi
-      (fun i _ ->
-        let s = Printf.sprintf "gen%d" i in
-        A.evar s)
-      l
-    |> A.pexp_tuple
-
-  let rec nest ~loc ~pair ~triple ~quad = function
-    | Quad (a, b, c, d) -> [%expr [%e quad] [%e a] [%e b] [%e c] [%e d]]
-    | Triple (a, b, c) -> [%expr [%e triple] [%e a] [%e b] [%e c]]
-    | Pair (a, b) ->
-        [%expr
-          [%e pair]
-            [%e nest ~loc ~pair ~triple ~quad a]
-            [%e nest ~loc ~pair ~triple ~quad b]]
-    | Elem a -> a
-
-  let to_gen ~loc t =
-    let pair = [%expr QCheck.Gen.pair] in
-    let triple = [%expr QCheck.Gen.triple] in
-    let quad = [%expr QCheck.Gen.quad] in
-    nest ~loc ~pair ~triple ~quad t
-
-  let to_obs ~loc t =
-    let pair = [%expr QCheck.Observable.pair] in
-    let triple = [%expr QCheck.Observable.triple] in
-    let quad = [%expr QCheck.Observable.quad] in
-    nest ~loc ~pair ~triple ~quad t
-
-  let to_pat ~loc t =
-    let fresh_id =
-      let id = ref 0 in
-      fun () ->
-        let x = !id in
-        let () = id := x + 1 in
-        Printf.sprintf "gen%d" x
-    in
-    let (module A) = Ast_builder.make loc in
-    let rec aux = function
-      | Quad (_, _, _, _) ->
-          let a = A.pvar @@ fresh_id () in
-          let b = A.pvar @@ fresh_id () in
-          let c = A.pvar @@ fresh_id () in
-          let d = A.pvar @@ fresh_id () in
-          [%pat? [%p a], [%p b], [%p c], [%p d]]
-      | Triple (_, _, _) ->
-          let a = A.pvar @@ fresh_id () in
-          let b = A.pvar @@ fresh_id () in
-          let c = A.pvar @@ fresh_id () in
-          [%pat? [%p a], [%p b], [%p c]]
-      | Pair (a, b) ->
-          let a = aux a in
-          let b = aux b in
-          [%pat? [%p a], [%p b]]
-      | Elem _ -> A.pvar @@ fresh_id ()
-    in
-    aux t
-end
+    [f] can be used to transform tuples, for instance:
+    {[
+    type t = Foo of int * int
+    ]}
 
-let map ~loc pat expr gen =
-  [%expr QCheck.Gen.map (fun [%p pat] -> [%e expr]) [%e gen]]
+    Without [f]:
+    {[
+    let gen = QCheck.Gen.(map (fun (x, y) -> (x, y)) (pair int int))
+    ]}
 
+    With [f], building Foo:
+    {[
+    let gen = QCheck.Gen.(map (fun (x, y) -> Foo (x, y)) (pair int int))
+    ]}
+*)
 let tuple ~loc ?(f = fun x -> x) tys =
```

Comment:
rename to `gen_tuple`?

Reply:
Done in d527aa353bf3f971791ff1a3e112d8d2d24179d1.